File: Webpage/statsPuller.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"net/http"

	"github.com/shirou/gopsutil/cpu"
	"github.com/shirou/gopsutil/disk"
	"github.com/shirou/gopsutil/mem"
)

// Middleware to handle CORS
func handleCORS(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*") // Allow all domains
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST") // Allowed HTTP methods
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization") // Allowed headers
		if r.Method == http.MethodOptions { // Pre-flight request
			w.WriteHeader(http.StatusOK)
			return
		}
		next.ServeHTTP(w, r)
	})
}

// Helper function to round a float to one decimal place
func roundToOneDecimalPlace(value float64) float64 {
	return math.Round(value*10) / 10
}

// Function to get system metrics (CPU, Memory, Disk)
func getMetrics(w http.ResponseWriter, r *http.Request) {
	// CPU usage
	cpuPercent, err := cpu.Percent(0, false)
	if err != nil {
		http.Error(w, "Error retrieving CPU usage", http.StatusInternalServerError)
		return
	}

	// Memory usage
	memStats, err := mem.VirtualMemory()
	if err != nil {
		http.Error(w, "Error retrieving memory usage", http.StatusInternalServerError)
		return
	}

	// Disk usage
	diskStats, err := disk.Usage("/")
	if err != nil {
		http.Error(w, "Error retrieving disk usage", http.StatusInternalServerError)
		return
	}

	// Create a structure to hold the data
	metrics := struct {
		CPU    float64 `json:"cpu"`
		Memory float64 `json:"memory"`
		Disk   float64 `json:"disk"`
	}{
		CPU:    roundToOneDecimalPlace(cpuPercent[0]),           // Get the first value from the CPU percentage array
		Memory: roundToOneDecimalPlace(memStats.UsedPercent),    // Get the percentage of memory used
		Disk:   roundToOneDecimalPlace(diskStats.UsedPercent),   // Get the percentage of disk space used
	}

	// Set the content type as JSON
	w.Header().Set("Content-Type", "application/json")
	// Write the response as JSON
	err = json.NewEncoder(w).Encode(metrics)
	if err != nil {
		http.Error(w, "Error encoding JSON response", http.StatusInternalServerError)
		return
	}
}

func main() {
	// Set up the server
	mux := http.NewServeMux()
	mux.HandleFunc("/metrics", getMetrics)

	// Wrap the mux with the CORS handler
	http.Handle("/", handleCORS(mux))

	// Start the server
	fmt.Println("Server started at http://localhost:8080")
	err := http.ListenAndServe(":8080", nil)
	if err != nil {
		log.Fatal("Error starting server:", err)
	}
}
// another test comment

File: Webpage/statsPuller.js
const express = require('express');
const cors = require('cors');
const si = require('systeminformation');
const rateLimit = require('express-rate-limit');

const app = express();
const PORT = process.env.PORT || 8080;

// Configure rate limiting
const limiter = rateLimit({
    windowMs: 1 * 60 * 1000, // 1 minute
    max: 60, // limit each IP to 60 requests per windowMs
    message: 'Too many requests from this IP, please try again later'
});

// Middleware
app.use(cors());
app.use(express.json());
app.use('/metrics', limiter);

// Helper function to round values to one decimal place
function roundToOneDecimalPlace(value) {
    return Math.round(value * 10) / 10;
}

// Helper function to convert bytes to GB
function bytesToGB(bytes) {
    return roundToOneDecimalPlace(bytes / (1024 ** 3));
}

// Function to fetch system metrics
async function getMetrics(req, res) {
    try {
        // Fetch all metrics in parallel for better performance
        const [cpuLoad, cpuInfo, memory, disks] = await Promise.all([
            si.currentLoad(),
            si.cpu(),
            si.mem(),
            si.fsSize()
        ]);

        // Calculate CPU metrics
        const cpuUsage = roundToOneDecimalPlace(cpuLoad.currentLoad);
        const totalCores = cpuInfo.cores;

        // Calculate memory metrics
        const memoryUsage = roundToOneDecimalPlace((memory.active / memory.total) * 100);
        const totalMemoryGB = bytesToGB(memory.total);
        const usedMemoryGB = bytesToGB(memory.active);

        // Calculate disk metrics
        let totalDiskSpace = 0;
        let totalUsedSpace = 0;

        disks.forEach(disk => {
            totalDiskSpace += disk.size;
            totalUsedSpace += disk.used;
        });

        const diskUsage = roundToOneDecimalPlace((totalUsedSpace / totalDiskSpace) * 100);
        const totalDiskSpaceGB = bytesToGB(totalDiskSpace);
        const totalUsedSpaceGB = bytesToGB(totalUsedSpace);

        // Prepare response object
        const metrics = {
            cpu: {
                usage: cpuUsage,
                cores: totalCores
            },
            memory: {
                usage: memoryUsage,
                total: totalMemoryGB,
                used: usedMemoryGB
            },
            disk: {
                usage: diskUsage,
                total: totalDiskSpaceGB,
                used: totalUsedSpaceGB
            },
            timestamp: new Date().toISOString()
        };

        res.json(metrics);
    } catch (error) {
        console.error("Error fetching system metrics:", error);
        res.status(500).json({
            error: "Failed to retrieve metrics",
            message: process.env.NODE_ENV === 'development' ? error.message : 'Internal server error'
        });
    }
}

// Routes
app.get('/metrics', getMetrics);

// Error handling middleware
app.use((err, req, res, next) => {
    console.error(err.stack);
    res.status(500).json({
        error: "Internal server error",
        message: process.env.NODE_ENV === 'development' ? err.message : 'Something went wrong'
    });
});

// Start server
app.listen(PORT, () => {
    console.log(`Server started at http://localhost:${PORT}`);
    console.log(`Environment: ${process.env.NODE_ENV || 'development'}`);
});